align/main.py

```diff
@@ -65,3 +66,24 @@ def schematic2layout(netlist_dir, pdk_dir, netlist_file=None, subckt=None, worki
                     mp.write(f'{true_stem} {true_stem}.gds\n')
                 elif file_.suffix == '.lef' and file_.stem != subckt:
                     lp.write(file_.read_text())
+        # Copy over necessary collateral & run PNR tool
+        pnr_dir = working_dir / '3_pnr'
+        pnr_dir.mkdir(exist_ok=True)
+        # TODO: Copying is bad ! Rewrite C++ code to accept fully qualified paths
+        (pnr_dir / (subckt + '.map')).write_text((primitive_dir / (subckt + '.map')).read_text())
+        (pnr_dir / (subckt + '.lef')).write_text((primitive_dir / (subckt + '.lef')).read_text())
+        (pnr_dir / (subckt + '.v')).write_text((topology_dir / (subckt + '.v')).read_text())
+        (pnr_dir / 'layers.json').write_text((pdk_dir / 'layers.json').read_text())
+        for file_ in topology_dir.iterdir():
+            if file_.suffix == '.const':
+                (pnr_dir / file_.name).write_text(file_.read_text())
+        for file_ in primitive_dir.iterdir():
+            if file_.suffix == '.json':
+                (pnr_dir / file_.name).write_text(file_.read_text())
+        generate_pnr(
+            pnr_dir,
+            (subckt + '.lef'),
```

Comment:
You are using f-strings sometimes and string addition other times. 
Missing newline (maybe we need to make sure the gitattrs is working now that the directories have been modified.

Reply:
Converged on fstrings. Thank you for pointing that out !